x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java

```diff
@@ -67,6 +100,95 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
+                if (originalCharacterQuoted) {
+                    if (quotingChar == '\\') {
+                        // in the original pattern, this is a single quoted character, add it to the partial string
+                        // that will be quoted in Java
+                        originalCharacterQuoted = false;
+                        lastTargetCharacterQuoted = true;
+                        partialQuotedString.append(c);
+                    } else if (c == quotingChar) {
+                        // the original pattern is closing the quoting,
+                        // do nothing for now, next character could open a new quoting block
+                        originalCharacterQuoted = false;
+                    } else {
+                        // any character that is not a quoting char is just added to the partial quoting string
+                        // because there could be more characters to quote after that
+                        partialQuotedString.append(c);
+                    }
+                } else {
+                    boolean characterProcessed = false;
+                    // the original pattern is not quoting
+                    if (MS_QUOTING_CHARS.contains(c)) {
+                        // next character(s) is quoted, start a quoted block on the target
+                        originalCharacterQuoted = true;
+                        lastTargetCharacterQuoted = true;
+                        quotingChar = c;
+                        characterProcessed = true;
+                    } else {
+                        // manage patterns that are different from MS to Java and have to be translated
+                        for (String[] item : MS_TO_JAVA_PATTERNS) {
+                            int fragmentLength = item[0].length();
+                            if (i + fragmentLength <= pattern.length() && item[0].equals(pattern.substring(i, i + fragmentLength))) {
+                                if (lastTargetCharacterQuoted) {
+                                    // now origin is not quoting for sure and the next block is a valid datetime pattern,
+                                    // that has to be translated and written as is (not quoted).
+                                    // Before doing this, let's flush the previously quoted string
+                                    // and quote it properly with Java syntax
+                                    lastTargetCharacterQuoted = false;
+                                    quoteAndAppend(result, partialQuotedString);
+                                    partialQuotedString = new StringBuilder();
+                                }
+                                // and then translate the pattern
+                                result.append(item[1]);
+                                characterProcessed = true;
+                                i += (fragmentLength - 1); // fast-forward, because the replaced pattern could be longer than one character
```

Comment:
Nit: the parentheses shouldn't be needed. Just in case any other changes will be needed.
Also, I'm generally in favour of using a while loop and explicitly incrementing by needed amount (vs. decrementing due to for).